### ISSUE TITLE
knit() cannot handle variable numbers of backticks

### ISSUE THREAD

**Issue by yutannihilation** [yihui/knitr#1621]
(Originally reported here: https://github.com/rstudio/rmarkdown/issues/1444#issuecomment-428504984)

`spin()` now uses various number of backticks (#1611), but currently `knit()` cannot handle them.

````` r
tmp_rmd <- tempfile(fileext = ".Rmd")
tmp_md <- tempfile(fileext = ".md")

writeLines(
  c('````{r}',
    'x <- "',
    '```',
    '"',
    '````'),
  con = tmp_rmd
)

cat(readLines(tmp_rmd), sep = "\n")
#> ````{r}
#> x <- "
#> ```
#> "
#> ````

res <- knitr::knit(tmp_rmd, output = tmp_md, quiet = TRUE)
cat(readLines(tmp_md), sep = "\n")
#> 
#> ```r
#> x <- "
#> #> Error: <text>:1:6: unexpected INCOMPLETE_STRING
#> #> 1: x <- "
#> #>          ^
#> ```
#> "
#> ````
`````

<sup>Created on 2018-10-10 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

To determine how many backticks are used to represent the end of a chunk, we need to look at the beginning of the chunk (in actual, the rule is [a bit more complex](https://spec.commonmark.org/0.28/#example-93)). But, in the current implementation of `split_file()`, begin chunks and end chunks are matched separately, so it seems difficult to fix without introducing a big change to the current implementation.

https://github.com/yihui/knitr/blob/3abb642649d186ba0a42d7e489ea3babd913b023/R/parser.R#L15-L16

**Comment by yihui** [comment]
I don't think this particular case will be too complicated to fix. You may take a look at `xfun::prose_index()`, which does a similar job. In particular, these lines: https://github.com/yihui/xfun/blob/d547831d/R/markdown.R#L17-L23

However, even with that fix, it still won't work for this case (four backticks in the string):

`````md
````{r}
x <- "
````
"
````
`````

**Comment by yutannihilation** [comment]
Oh, thanks. Maybe I thought too much.

That case doesn't seem supposed to work because it's not a valid markdown...?

**Comment by yihui** [comment]
Well, it is valid Markdown: it is a code block with `x <- "`, followed by `"`, and another four backticks (will be silently ignored since the fence is not closed). You may say it is invalid R Markdown because the second series of four backticks closed the code chunk too early. The R Markdown parser in **knitr** is based on regular expressions, and cannot be so smart to know the second four backticks is actually a part of the R code.

**Comment by yutannihilation** [comment]
Sorry, this is valid as Markdown / R Markdown. What I meant to say was `x <- "` is not a valid R code.

> The R Markdown parser in knitr is based on regular expressions, and cannot be so smart to know the second four backticks is actually a part of the R code.

Sorry, I don't get your point. Do you think knitr's parser should parse this as a code block of the below instead of `x <- "`?

````` r
x <- "
````
"
`````

**Comment by yihui** [comment]
Yes. Well, maybe. I don't really know. Perhaps it should just fail. There isn't a right answer here. Users could write an R code chunk with incomplete R code, e.g.,

````
```{r, error=TRUE}
x <- "
```
````

And the output will be

![image](https://user-images.githubusercontent.com/163582/54072917-b139a600-4246-11e9-9bbb-c216bf997e0b.png)

So I guess we can ignore the edge case I first mentioned, and only fix the original issue you brought forward.

**Comment by yutannihilation** [comment]
I believe it should fail; R Markdown itself should not be aware of code inside chunks because the code is not always written in R.

That said, after some thinking, now I feel you are right in that it's  nice if knitr can detect these mistakes and raise a helpful error message :)

**Comment by atusy** [comment]
I tried to solve this problem.
The result gives no error, but does not properly enclose the code block.
Now I see this issue requires a drastic change so that output code block has as many backticks as the input chunk.


````
```r
x <- "
```
"
```
````

https://github.com/atusy/knitr/tree/1621

**Comment by yihui** [comment]
The fix https://github.com/yihui/knitr/compare/master...atusy:1621 seems to be on the right track to me. I didn't read it carefully or test it.

> Now I see this issue requires a drastic change so that output code block has as many backticks as the input chunk.

I'm not sure what that means, but the output blocks will intelligently use N+1 backticks when N backticks are detected in the output: https://github.com/yihui/knitr/blob/7d2dd40bfaad6df54cd451d6553c42f32a34d56f/R/hooks-md.R#L157-L162 Will a trick like that help?

**Comment by atusy** [comment]
@yihui Oh thanks for hint. I'll see if I can implement simply with your trick (maybe next weekend)

**Comment by yutannihilation** [comment]
Thanks for fixing!!! 🎉
